core/modules/replica/Request.h

```diff
@@ -0,0 +1,398 @@
+     *    which initiated the abort to take care of putting the object into
+     *    a proper state.
+     */
+    bool isAborted(boost::system::error_code const& ec) const;
+
+    /**
+     * Ensure the object is in the deseride internal state. Throw an
+     * exception otherwise.
+     *
+     * NOTES: normally this condition should never been seen unless
+     *        there is a problem with the application implementation
+     *        or the underlying run-time system.
+     *
+     * @throws std::logic_error
+     */
+    void assertState(State desiredState) const;
+
+    /**
+     * Set the desired primary and extended state.
+     *
+     * The change of the state is done via a method to allow extra actions
+     * at this step, such as:
+     *
+     * - reporting change state in a debug stream
+     * - verifying the correctness of the state transition
+     */
+    void setState(State state,
+                  ExtendedState extendedStat);
+
+protected:
+
+    // Parameters of the object
+
+    ServiceProvider::pointer _serviceProvider;
+
+    std::string _type;
+    std::string _id;                    ///< own identifier
+    std::string _duplicateRequestId;    ///< effective identifier of a remote (worker-side) request where applies
+    std::string _worker;
+
+    int  _priority;
+    bool _keepTracking;
+    bool _allowDuplicate;
+
+    // Primary and extended states of the request
+
+    State         _state;
+    ExtendedState _extendedState;
+
+    ExtendedCompletionStatus _extendedServerStatus;
+
+    /// Performance counters
+    Performance _performance;
+
+    /// Buffer for data moved over the network
+    std::shared_ptr<ProtocolBuffer> _bufferPtr;
+
+    // Parameters of a worker obtained from a configuration
+    WorkerInfo const& _workerInfo;
+
+    /// This time is used to in the communication protocol for requests
+    /// which may require multiple retries or any time spacing between network
+    /// operation.
+    ///
+    /// The current class doesn't manager this time. It's here just because
+    /// it's required by virtually all implementations of requests. And it's
+    /// up to subclasses to manage this timer.
+    ///
+    /// The only role of this class (apart from providing the objects)
+    /// is to set up the default timer interval from a condifuration.
+    unsigned int                _timerIvalSec;
+    boost::asio::deadline_timer _timer;
+
+    /// This timer is used (if configured) to limit the total run time
+    /// of a request. The timer starts when the request is started. And it's
+    /// explicitly finished when a request finishes (successfully or not).
+    ///
+    /// If the time has a chance to expire then the request would finish
+    /// with status: FINISHED::EXPIRED.
+    unsigned int                _requestExpirationIvalSec;
+    boost::asio::deadline_timer _requestExpirationTimer;
+
+    /// Mutex guarding internal state
+    mutable std::mutex _mtx;

```

Comment:
guarding what state exactly - is it just `_state` and `_extendedState`? Or more internal variables than that?
There are times when the state is tested when the mutex is not held by a lock, it seems dangerous?

Reply:
For the record, this concern (I hope) was cleared during a dedicated in-person meeting. Please, let me know if any further discussion is needed here.